### PR TITLE
design: Propose a MetalLB operator

### PR DESCRIPTION
**Pull request by russellb**
This proposes creating a new repository under the metallb github
organization to allow interested parties to collaborate on an operator
as an alternative method for managing the deployment of MetalLB on a
cluster.

### CODE REVIEW

File: design/metallb-operator.md
Comment (fedepaol) [inline design/metallb-operator.md:26]: +1 on this. The only caveat is having a good plan if we consider moving the "experimental" CRD over to metallb itself.
I would aim at having the same configuration CRD, and in case moving it as it is to metallb if / when it gets accepted.
This would avoid the need of supporting both apis imo

Comment (russellb) [reply design/metallb-operator.md:26]: Yeah, that would be the hope ... that we end up with something that's working well enough that we consider migrating it over to MetalLB.  if we need to take a fresh look at it and a v2 of the CRDs, we'd hopefully implement an automated migration if technically feasible.  It's hard to predict, but nobody wants to make our own job harder than necessary, and that includes inflicting pain on our users. :)

**Review by daxmc99 — APPROVED** [review]
👍 This sounds like a great idea and I support moving forward on it.

File: design/metallb-operator.md
Comment (msherif1234) [reply design/metallb-operator.md:26]: that is actually the current plan since the functionality scope of this config CRD will be limited we will be taking few  parts that are more or less agreed on from #850 and use it here

File: design/metallb-operator.md
Comment (rata) [inline design/metallb-operator.md:33]: Do we need some automation added to the releases of "core" MetalLB? Or to releases of the metallb operator?

**Review by rata — APPROVED** [review]
LGTM, thanks! @russellb feel free to ping me on slack if we need to coordinate any of this :)

File: design/metallb-operator.md
Comment (russellb) [reply design/metallb-operator.md:33]: I don't know yet.  :)  Probably!  I could add another note to this list about considering release automation.

**Comment by russellb** [comment]
Thanks for the approvals.  I figure the new repo should get an ack from all the maintainers, so I'll wait on @johananl as well.

@rata Thanks for the coordination help.  I think I've got the permissions I need to get the repo created and moving.  We'd use the same meeting we have to provide updates on how it goes.

**Comment by fedepaol** [comment]
Just to add that we already started polishing a prototype of the operator under my personal gh account, it would be ideal to migrate that once we reach consensus (https://github.com/fedepaol/metallboperator ).

**Comment by russellb** [comment]
> Just to add that we already started polishing a prototype of the operator under my personal gh account, it would be ideal to migrate that once we reach consensus (https://github.com/fedepaol/metallboperator ).

Thanks!  Once this PR is settled, I'll look over what you have and we'll sort out the best way to get your existing code in.

File: design/metallb-operator.md
Comment (russellb) [inline design/metallb-operator.md:51]: Based on @fedepaol 's comment that he has a repo where they've kicked off this work already, it probably makes sense to add him as a committer after I complete a review of what's done so far.

File: design/metallb-operator.md
Comment (johananl) [inline design/metallb-operator.md:26]:
```suggestion
have to disrupt the core MetalLB project.
```

File: design/metallb-operator.md
Comment (johananl) [inline design/metallb-operator.md:33]:
```suggestion
- Publish the operator on operatorhub.io to make it discoverable.
```

File: design/metallb-operator.md
Comment (johananl) [inline design/metallb-operator.md:34]: Let's say we find a good CRD data model in the new repo - what would we do then? "Port" the CRD code to the MetalLB repo and get rid of the operator?

File: design/metallb-operator.md
Comment (johananl) [inline design/metallb-operator.md:41]:
```suggestion
  interface in the core MetalLB project, but instead provide a place to
```

File: design/metallb-operator.md
Comment (johananl) [inline design/metallb-operator.md:47]:
```suggestion
* Create github.com/metallb/metallb-operator.
```

**Review by johananl — APPROVED** [review]
Interesting idea.

FYI, Equinix Metal wrote an operator which, among the rest, [manages MetalLB](https://github.com/equinix/cloud-provider-equinix-metal/tree/master/metal/loadbalancers/metallb). That project is published under the Apache license. This could be useful for code reuse and/or inspiration.

Added a few nits. Otherwise, LGTM.

File: design/metallb-operator.md
Comment (russellb) [reply design/metallb-operator.md:34]: Yes, I imagine we could move the CRD handling code to MetalLB.  I imagine the operator would remain, though.  It would still be responsible for managing the controller deployment and the speaker daemonset.  It would do the equivalent of what helm does.

**Comment by russellb** [comment]
Thanks for the feedback, everyone!  I'm going to merge this, but please continue to share feedback if you think of any other comments or questions.  Leave comments here or ping me on slack.

**Pull request merged**
